File: VillagesTown/Engines/RecruitmentEngine.swift
//
//  RecruitmentEngine.swift
//  VillagesTown
//
//  Created by Claude Code
//

import Foundation

class RecruitmentEngine {

    // MARK: - Recruitment
    func canRecruit(unitType: Unit.UnitType, quantity: Int, in village: Village) -> (can: Bool, reason: String) {
        // Check if village has required military building
        let requiredBuilding = getRequiredBuilding(for: unitType)

        if !village.buildings.contains(where: { $0.name == requiredBuilding }) {
            return (false, "Requires \(requiredBuilding)")
        }

        // Check population for recruitment
        if village.population < quantity {
            return (false, "Not enough population. Need \(quantity), have \(village.population)")
        }

        // Get unit stats
        let stats = Unit.getStats(for: unitType)
        let totalCost = stats.cost.mapValues { $0 * quantity }

        // Check global resources
        let globalResources = GameManager.shared.getGlobalResources(playerID: village.owner)
        for (resource, amount) in totalCost {
            let available = globalResources[resource] ?? 0
            if available < amount {
                return (false, "Not enough \(resource.name). Need \(amount), have \(available)")
            }
        }

        return (true, "Can recruit")
    }

    func recruitUnits(unitType: Unit.UnitType, quantity: Int, in village: inout Village, at coordinates: CGPoint) -> [Unit] {
        let check = canRecruit(unitType: unitType, quantity: quantity, in: village)

        guard check.can else {
            print("❌ \(village.name): Cannot recruit \(unitType.rawValue) - \(check.reason)")
            return []
        }

        // Get unit stats and costs
        let stats = Unit.getStats(for: unitType)
        let totalCost = stats.cost.mapValues { $0 * quantity }

        // Pay costs from global pool
        guard GameManager.shared.spendResources(playerID: village.owner, cost: totalCost) else {
            print("❌ \(village.name): Failed to spend resources")
            return []
        }

        // Reduce population
        village.modifyPopulation(by: -quantity)

        // Find empty adjacent tile for spawning units
        let spawnLocation = GameManager.shared.map.findEmptyAdjacentTile(to: coordinates)

        // Create units
        var units: [Unit] = []
        for _ in 0..<quantity {
            let unit = Unit(type: unitType, owner: village.owner, coordinates: spawnLocation)
            units.append(unit)
        }

        print("⚔️ \(village.name): Recruited \(quantity) \(stats.name) at (\(Int(spawnLocation.x)), \(Int(spawnLocation.y)))!")
        return units
    }

    func getRequiredBuilding(for unitType: Unit.UnitType) -> String {
        switch unitType.category {
        case "Infantry":
            return "Barracks"
        case "Ranged":
            return "Archery Range"
        default:
            return "Barracks"
        }
    }

    func getAvailableUnits(for village: Village) -> [Unit.UnitType] {
        var available: [Unit.UnitType] = []

        // Check which military buildings exist
        let hasBarracks = village.buildings.contains(where: { $0.name == "Barracks" })
        let hasArcheryRange = village.buildings.contains(where: { $0.name == "Archery Range" })

        // Infantry
        if hasBarracks {
            available.append(contentsOf: [.militia, .spearman, .swordsman])
        }

        // Ranged
        if hasArcheryRange {
            available.append(contentsOf: [.archer, .crossbowman])
        }

        return available
    }
}

File: lib/engines/recruitment_engine.dart
import '../data/models/game_modifier.dart';
import '../data/models/geo_coordinate.dart';
import '../data/models/resource.dart';
import '../data/models/unit.dart';
import '../data/models/unit_type.dart';
import '../data/models/village.dart';
import 'game_manager.dart';

class RecruitmentEngine {
  static const _peasantUnits = {UnitType.militia, UnitType.spearman};

  (bool can, String reason) canRecruit(UnitType unitType, int quantity, Village village) {
    final game = GameManager.shared;

    // Peasant War: only militia and spearmen
    if (game.activeModifiers.contains(GameModifier.peasantWar) && !_peasantUnits.contains(unitType)) {
      return (false, 'Only militia and spearmen (Peasant War)');
    }

    // Check mobilization cap
    if (village.recruitsThisTurn + quantity > village.maxRecruitsPerTurn) {
      return (false, 'Mobilization cap reached');
    }

    // Check required building
    final requiredBuilding = getRequiredBuilding(unitType);
    if (requiredBuilding != null && !village.buildings.any((b) => b.name == requiredBuilding)) {
      return (false, 'Requires $requiredBuilding');
    }

    // Check population
    if (village.population < quantity * 10) {
      return (false, 'Insufficient population');
    }

    // Check cost (with Gold Standard modifier)
    final stats = unitType.stats;
    var totalCost = stats.cost.map((k, v) => MapEntry(k, v * quantity));
    if (game.activeModifiers.contains(GameModifier.goldStandard)) {
      totalCost = Map.from(totalCost);
      if (totalCost.containsKey(Resource.gold)) {
        totalCost[Resource.gold] = totalCost[Resource.gold]! * 2;
      }
    }
    if (!game.canAfford(village.owner, totalCost)) {
      return (false, 'Insufficient resources');
    }

    return (true, '');
  }

  List<Unit> recruitUnits(UnitType unitType, int quantity, Village village, GeoCoordinate coordinates) {
    final (can, _) = canRecruit(unitType, quantity, village);
    if (!can) return [];

    final game = GameManager.shared;
    final stats = unitType.stats;
    var totalCost = stats.cost.map((k, v) => MapEntry(k, v * quantity));
    if (game.activeModifiers.contains(GameModifier.goldStandard)) {
      totalCost = Map.from(totalCost);
      if (totalCost.containsKey(Resource.gold)) {
        totalCost[Resource.gold] = totalCost[Resource.gold]! * 2;
      }
    }

    if (!game.spendResources(village.owner, totalCost)) return [];

    // Reduce population
    village.modifyPopulation(-quantity * 10);
    village.recruitsThisTurn += quantity;

    // Get building levels for unit bonuses
    final barracks = village.buildings.where((b) => b.name == 'Barracks').toList();
    final archeryRange = village.buildings.where((b) => b.name == 'Archery Range').toList();
    final stables = village.buildings.where((b) => b.name == 'Stables').toList();

    final barracksLevel = barracks.isNotEmpty ? barracks.first.level : 0;
    final archeryRangeLevel = archeryRange.isNotEmpty ? archeryRange.first.level : 0;
    final stablesLevel = stables.isNotEmpty ? stables.first.level : 0;

    // Military hub bonus: +15% attack and defense on recruited units
    final isMilitaryHub = village.specialization == VillageSpecialization.militaryHub;

    // Create units with building bonuses
    final units = <Unit>[];
    for (var i = 0; i < quantity; i++) {
      final unit = Unit.create(
        unitType,
        village.owner,
        coordinates,
        barracksLevel: barracksLevel,
        archeryRangeLevel: archeryRangeLevel,
        stablesLevel: stablesLevel,
      );
      if (isMilitaryHub) {
        unit.bonusAttack += (unitType.stats.attack * 0.15).round();
        unit.bonusDefense += (unitType.stats.defense * 0.15).round();
      }
      units.add(unit);
    }

    // Add to army at village (owned by same player!)
    final armiesAtVillage = game.getArmiesAt(village.id)
        .where((a) => a.owner == village.owner)
        .toList();
    if (armiesAtVillage.isNotEmpty) {
      final army = armiesAtVillage.first;
      army.addUnits(units);
      game.updateArmy(army);
    } else {
      game.createArmy(units, village.id, village.owner);
    }

    game.updateVillage(village);
    return units;
  }

  String? getRequiredBuilding(UnitType unitType) {
    return switch (unitType) {
      UnitType.militia || UnitType.spearman || UnitType.swordsman => 'Barracks',
      UnitType.archer || UnitType.crossbowman => 'Archery Range',
      UnitType.lightCavalry || UnitType.knight => 'Stables',
    };
  }

  List<UnitType> getAvailableUnits(Village village) {
    final game = GameManager.shared;
    final units = <UnitType>[];
    final hasBarracks = village.buildings.any((b) => b.name == 'Barracks');
    final hasArchery = village.buildings.any((b) => b.name == 'Archery Range');
    final hasStables = village.buildings.any((b) => b.name == 'Stables');

    if (hasBarracks) {
      units.addAll([UnitType.militia, UnitType.spearman, UnitType.swordsman]);
    }
    if (hasArchery) {
      units.addAll([UnitType.archer, UnitType.crossbowman]);
    }
    if (hasStables) {
      units.addAll([UnitType.lightCavalry, UnitType.knight]);
    }

    // Peasant War: filter to only militia and spearmen
    if (game.activeModifiers.contains(GameModifier.peasantWar)) {
      units.removeWhere((u) => !_peasantUnits.contains(u));
    }

    return units;
  }
}
